sklearn/preprocessing/_discretization.py

```diff
@@ -177,6 +183,14 @@ def fit(self, X, y=None):
                 bin_edges[jj] = (centers[1:] + centers[:-1]) * 0.5
                 bin_edges[jj] = np.r_[col_min, bin_edges[jj], col_max]
 
+            # Remove redundant bins (i.e., bins whose width = 0)
+            if self.strategy in ('quantile', 'kmeans'):
+                bin_edges[jj] = np.unique(bin_edges[jj])
+                if len(bin_edges[jj]) - 1 != n_bins[jj]:
+                    warnings.warn('Redundant bins (i.e., bins whose width = 0)'
```

Comment:
This means the user would get a `UserWarning` without being able or knowing how to fix it. We usually try to tell the user how to handle and remove the warnings, don't we?

Reply:
We've solved the problem for the user. We've already raised similar warning when certain feature is constant.

Reply:
Then we can add something like "Setting the number of bins for feature %d to %d."

Reply:
> Then we can add something like "Setting the number of bins for feature %d to %d."

I don't think this will solve the problem when data is skewed.